source/material_model/averaging.cc

```diff
@@ -0,0 +1,412 @@
+                  double weight = alfad * ((1 + 3 * (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)));
+                  if (distance / h > 1)
+                    weight = 0;
+                  sum_value += weight / values_out[j];
+                  sum_weights += weight;
+                }
+              const double average = sum_weights / sum_value;
+
+              temp_values[i] = average;
+            };
+          for (unsigned int i = 0; i<N; ++i)
+            {
+              values_out[i] = temp_values[i];
+            }
+        }
+      else if (averaging_operation_lower == "nwd geometric average")
+        {
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
+              double h = bell_shape_limit;
+              double alfad = 0;
+              if (dim==2)
+                {
+                  alfad = 5/(numbers::PI * h * h);
+                }
+              else
+                {
+                  alfad = 106/(16 * numbers::PI * h * h * h);
+                }
+
+              Assert (values_out[i] >= 0,
+                      ExcMessage ("Computing the geometric average "
+                                  "only makes sense for non-negative "
+                                  "quantities."));
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  if (position[i].distance(position[j]) > max_distance)
+                    max_distance = position[i].distance(position[j]);
+                }
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  double distance = position[i].distance(position[j])/max_distance;
+                  double weight = alfad * ((1 + 3 * (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)));
+                  if (distance / h > 1)
+                    weight = 0;
+                  sum_value += weight*log(values_out[j]);
+                  sum_weights += weight;
+                }
+
+              const double average = std::exp (sum_value/sum_weights);
+
+              temp_values[i] = average;
+            };
+          for (unsigned int i = 0; i<N; ++i)
+            {
+              values_out[i] = temp_values[i];
+            }
+          //break;
+        }
+      else
+        {
+          AssertThrow (false,
+                       ExcMessage ("This averaging operation is not implemented."));
+        }
+    }
+
+    template <int dim>
+    void
+    Averaging<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                             typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      base_model -> evaluate(in,out);
+      if (out.viscosities.size() > 1)
```

Comment:
you look at the viscosities here as a proxy for the sizes of the other arrays, right? If so, maybe say so in a comment.